File: ontos/src/backend/src/models/assets.py
"""
Unified Asset Models

This module provides platform-agnostic asset type definitions and metadata models
for the pluggable connector architecture. It enables Datasets and Contracts to
reference any asset type uniformly across Unity Catalog, Snowflake, Kafka, PowerBI,
and other platforms.
"""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, Field


# ============================================================================
# Asset Categories
# ============================================================================

class AssetCategory(str, Enum):
    """High-level categorization of assets by their primary function."""
    DATA = "data"           # Tables, views, topics, streams
    COMPUTE = "compute"     # Functions, models, jobs, notebooks
    SEMANTIC = "semantic"   # Metrics, measures, dimensions, KPIs
    VISUALIZATION = "viz"   # Dashboards, reports, charts
    STORAGE = "storage"     # Files, objects, volumes
    OTHER = "other"         # Catch-all for uncategorized assets


# ============================================================================
# Unified Asset Types
# ============================================================================

class UnifiedAssetType(str, Enum):
    """
    Unified enumeration of all supported asset types across platforms.
    
    Naming convention: {PLATFORM}_{ASSET_TYPE}
    This allows platform-specific handling while maintaining a unified interface.
    """
    # -------------------------------------------------------------------------
    # Unity Catalog / Databricks - Data Assets
    # -------------------------------------------------------------------------
    UC_TABLE = "uc_table"
    UC_VIEW = "uc_view"
    UC_MATERIALIZED_VIEW = "uc_materialized_view"
    UC_STREAMING_TABLE = "uc_streaming_table"
    UC_VOLUME = "uc_volume"
    UC_VOLUME_FILE = "uc_volume_file"
    
    # -------------------------------------------------------------------------
    # Unity Catalog / Databricks - Compute Assets
    # -------------------------------------------------------------------------
    UC_FUNCTION = "uc_function"
    UC_MODEL = "uc_model"
    UC_NOTEBOOK = "uc_notebook"
    UC_JOB = "uc_job"
    UC_PIPELINE = "uc_pipeline"  # DLT / Spark Declarative Pipelines (SDP)
    
    # -------------------------------------------------------------------------
    # Unity Catalog / Databricks - Semantic Assets (NEW)
    # -------------------------------------------------------------------------
    UC_METRIC = "uc_metric"  # Unity Catalog metrics (AI/BI)
    
    # -------------------------------------------------------------------------
    # Snowflake - Data Assets
    # -------------------------------------------------------------------------
    SNOWFLAKE_TABLE = "snowflake_table"
    SNOWFLAKE_VIEW = "snowflake_view"
    SNOWFLAKE_MATERIALIZED_VIEW = "snowflake_materialized_view"
    SNOWFLAKE_STREAM = "snowflake_stream"
    SNOWFLAKE_STAGE = "snowflake_stage"
    
    # -------------------------------------------------------------------------
    # Snowflake - Compute Assets
    # -------------------------------------------------------------------------
    SNOWFLAKE_FUNCTION = "snowflake_function"
    SNOWFLAKE_PROCEDURE = "snowflake_procedure"
    SNOWFLAKE_TASK = "snowflake_task"
    
    # -------------------------------------------------------------------------
    # Kafka - Data Assets
    # -------------------------------------------------------------------------
    KAFKA_TOPIC = "kafka_topic"
    KAFKA_SCHEMA = "kafka_schema"  # Schema Registry schemas
    
    # -------------------------------------------------------------------------
    # PowerBI - Visualization & Semantic Assets
    # -------------------------------------------------------------------------
    POWERBI_DATASET = "powerbi_dataset"
    POWERBI_SEMANTIC_MODEL = "powerbi_semantic_model"
    POWERBI_DASHBOARD = "powerbi_dashboard"
    POWERBI_REPORT = "powerbi_report"
    POWERBI_DATAFLOW = "powerbi_dataflow"
    
    # -------------------------------------------------------------------------
    # Cloud Storage - Storage Assets
    # -------------------------------------------------------------------------
    S3_BUCKET = "s3_bucket"
    S3_OBJECT = "s3_object"
    ADLS_CONTAINER = "adls_container"
    ADLS_PATH = "adls_path"
    GCS_BUCKET = "gcs_bucket"
    GCS_OBJECT = "gcs_object"
    
    # -------------------------------------------------------------------------
    # API & External - Other Assets
    # -------------------------------------------------------------------------
    EXTERNAL_API = "external_api"
    EXTERNAL_DATABASE = "external_database"
    
    # -------------------------------------------------------------------------
    # Generic / Unknown
    # -------------------------------------------------------------------------
    GENERIC = "generic"
    UNKNOWN = "unknown"


# ============================================================================
# Asset Type Metadata
# ============================================================================

# Mapping of asset types to their categories
ASSET_TYPE_CATEGORIES: Dict[UnifiedAssetType, AssetCategory] = {
    # UC Data
    UnifiedAssetType.UC_TABLE: AssetCategory.DATA,
    UnifiedAssetType.UC_VIEW: AssetCategory.DATA,
    UnifiedAssetType.UC_MATERIALIZED_VIEW: AssetCategory.DATA,
    UnifiedAssetType.UC_STREAMING_TABLE: AssetCategory.DATA,
    UnifiedAssetType.UC_VOLUME: AssetCategory.STORAGE,
    UnifiedAssetType.UC_VOLUME_FILE: AssetCategory.STORAGE,
    # UC Compute
    UnifiedAssetType.UC_FUNCTION: AssetCategory.COMPUTE,
    UnifiedAssetType.UC_MODEL: AssetCategory.COMPUTE,
    UnifiedAssetType.UC_NOTEBOOK: AssetCategory.COMPUTE,
    UnifiedAssetType.UC_JOB: AssetCategory.COMPUTE,
    UnifiedAssetType.UC_PIPELINE: AssetCategory.COMPUTE,
    # UC Semantic
    UnifiedAssetType.UC_METRIC: AssetCategory.SEMANTIC,
    # Snowflake Data
    UnifiedAssetType.SNOWFLAKE_TABLE: AssetCategory.DATA,
    UnifiedAssetType.SNOWFLAKE_VIEW: AssetCategory.DATA,
    UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW: AssetCategory.DATA,
    UnifiedAssetType.SNOWFLAKE_STREAM: AssetCategory.DATA,
    UnifiedAssetType.SNOWFLAKE_STAGE: AssetCategory.STORAGE,
    # Snowflake Compute
    UnifiedAssetType.SNOWFLAKE_FUNCTION: AssetCategory.COMPUTE,
    UnifiedAssetType.SNOWFLAKE_PROCEDURE: AssetCategory.COMPUTE,
    UnifiedAssetType.SNOWFLAKE_TASK: AssetCategory.COMPUTE,
    # Kafka
    UnifiedAssetType.KAFKA_TOPIC: AssetCategory.DATA,
    UnifiedAssetType.KAFKA_SCHEMA: AssetCategory.DATA,
    # PowerBI
    UnifiedAssetType.POWERBI_DATASET: AssetCategory.DATA,
    UnifiedAssetType.POWERBI_SEMANTIC_MODEL: AssetCategory.SEMANTIC,
    UnifiedAssetType.POWERBI_DASHBOARD: AssetCategory.VISUALIZATION,
    UnifiedAssetType.POWERBI_REPORT: AssetCategory.VISUALIZATION,
    UnifiedAssetType.POWERBI_DATAFLOW: AssetCategory.DATA,
    # Cloud Storage
    UnifiedAssetType.S3_BUCKET: AssetCategory.STORAGE,
    UnifiedAssetType.S3_OBJECT: AssetCategory.STORAGE,
    UnifiedAssetType.ADLS_CONTAINER: AssetCategory.STORAGE,
    UnifiedAssetType.ADLS_PATH: AssetCategory.STORAGE,
    UnifiedAssetType.GCS_BUCKET: AssetCategory.STORAGE,
    UnifiedAssetType.GCS_OBJECT: AssetCategory.STORAGE,
    # External
    UnifiedAssetType.EXTERNAL_API: AssetCategory.OTHER,
    UnifiedAssetType.EXTERNAL_DATABASE: AssetCategory.DATA,
    # Generic
    UnifiedAssetType.GENERIC: AssetCategory.OTHER,
    UnifiedAssetType.UNKNOWN: AssetCategory.OTHER,
}

# Mapping of asset types that support schema (columns/fields)
SCHEMA_SUPPORTING_TYPES: set = {
    UnifiedAssetType.UC_TABLE,
    UnifiedAssetType.UC_VIEW,
    UnifiedAssetType.UC_MATERIALIZED_VIEW,
    UnifiedAssetType.UC_STREAMING_TABLE,
    UnifiedAssetType.SNOWFLAKE_TABLE,
    UnifiedAssetType.SNOWFLAKE_VIEW,
    UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW,
    UnifiedAssetType.KAFKA_TOPIC,  # If using Schema Registry
    UnifiedAssetType.POWERBI_DATASET,
    UnifiedAssetType.POWERBI_SEMANTIC_MODEL,
}

# Mapping of connector types to supported asset types
CONNECTOR_ASSET_TYPES: Dict[str, List[UnifiedAssetType]] = {
    "databricks": [
        UnifiedAssetType.UC_TABLE,
        UnifiedAssetType.UC_VIEW,
        UnifiedAssetType.UC_MATERIALIZED_VIEW,
        UnifiedAssetType.UC_STREAMING_TABLE,
        UnifiedAssetType.UC_VOLUME,
        UnifiedAssetType.UC_VOLUME_FILE,
        UnifiedAssetType.UC_FUNCTION,
        UnifiedAssetType.UC_MODEL,
        UnifiedAssetType.UC_NOTEBOOK,
        UnifiedAssetType.UC_JOB,
        UnifiedAssetType.UC_PIPELINE,
        UnifiedAssetType.UC_METRIC,
    ],
    "snowflake": [
        UnifiedAssetType.SNOWFLAKE_TABLE,
        UnifiedAssetType.SNOWFLAKE_VIEW,
        UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW,
        UnifiedAssetType.SNOWFLAKE_STREAM,
        UnifiedAssetType.SNOWFLAKE_STAGE,
        UnifiedAssetType.SNOWFLAKE_FUNCTION,
        UnifiedAssetType.SNOWFLAKE_PROCEDURE,
        UnifiedAssetType.SNOWFLAKE_TASK,
    ],
    "kafka": [
        UnifiedAssetType.KAFKA_TOPIC,
        UnifiedAssetType.KAFKA_SCHEMA,
    ],
    "powerbi": [
        UnifiedAssetType.POWERBI_DATASET,
        UnifiedAssetType.POWERBI_SEMANTIC_MODEL,
        UnifiedAssetType.POWERBI_DASHBOARD,
        UnifiedAssetType.POWERBI_REPORT,
        UnifiedAssetType.POWERBI_DATAFLOW,
    ],
    "s3": [
        UnifiedAssetType.S3_BUCKET,
        UnifiedAssetType.S3_OBJECT,
    ],
}


def get_asset_category(asset_type: UnifiedAssetType) -> AssetCategory:
    """Get the category for a given asset type."""
    return ASSET_TYPE_CATEGORIES.get(asset_type, AssetCategory.OTHER)


def supports_schema(asset_type: UnifiedAssetType) -> bool:
    """Check if an asset type supports schema (columns/fields)."""
    return asset_type in SCHEMA_SUPPORTING_TYPES


def get_connector_type_for_asset(asset_type: UnifiedAssetType) -> Optional[str]:
    """Get the connector type that handles a given asset type."""
    for connector_type, asset_types in CONNECTOR_ASSET_TYPES.items():
        if asset_type in asset_types:
            return connector_type
    return None


# ============================================================================
# Schema Models (for assets that have columnar structure)
# ============================================================================

class ColumnInfo(BaseModel):
    """Information about a single column/field in an asset's schema."""
    name: str = Field(..., description="Column name")
    data_type: str = Field(..., description="Data type (platform-specific)")
    logical_type: Optional[str] = Field(None, description="Logical/normalized type")
    nullable: bool = Field(True, description="Whether the column allows nulls")
    description: Optional[str] = Field(None, description="Column description/comment")
    is_primary_key: bool = Field(False, description="Whether this is part of the primary key")
    is_partition_key: bool = Field(False, description="Whether this is a partition column")
    default_value: Optional[str] = Field(None, description="Default value if any")
    
    # Extended metadata
    tags: Dict[str, str] = Field(default_factory=dict, description="Column-level tags")
    properties: Dict[str, Any] = Field(default_factory=dict, description="Additional properties")
    
    model_config = {"from_attributes": True}


class SchemaInfo(BaseModel):
    """Schema information for assets that have columnar structure."""
    columns: List[ColumnInfo] = Field(default_factory=list, description="List of columns")
    primary_key: Optional[List[str]] = Field(None, description="Primary key column names")
    partition_columns: Optional[List[str]] = Field(None, description="Partition column names")
    clustering_columns: Optional[List[str]] = Field(None, description="Clustering column names")
    
    # Schema versioning
    schema_version: Optional[str] = Field(None, description="Schema version if tracked")
    
    model_config = {"from_attributes": True}
    
    @property
    def column_count(self) -> int:
        """Return the number of columns."""
        return len(self.columns)
    
    def get_column(self, name: str) -> Optional[ColumnInfo]:
        """Get a column by name (case-insensitive)."""
        name_lower = name.lower()
        for col in self.columns:
            if col.name.lower() == name_lower:
                return col
        return None


# ============================================================================
# Asset Metadata Models
# ============================================================================

class AssetOwnership(BaseModel):
    """Ownership information for an asset."""
    owner: Optional[str] = Field(None, description="Primary owner (user or group)")
    owner_email: Optional[str] = Field(None, description="Owner email")
    steward: Optional[str] = Field(None, description="Data steward")
    team: Optional[str] = Field(None, description="Owning team")
    
    model_config = {"from_attributes": True}


class AssetStatistics(BaseModel):
    """Statistics about an asset (when available)."""
    row_count: Optional[int] = Field(None, description="Number of rows/records")
    size_bytes: Optional[int] = Field(None, description="Size in bytes")
    partition_count: Optional[int] = Field(None, description="Number of partitions")
    file_count: Optional[int] = Field(None, description="Number of files (for table formats)")
    last_modified: Optional[datetime] = Field(None, description="Last modification timestamp")
    last_accessed: Optional[datetime] = Field(None, description="Last access timestamp")
    
    model_config = {"from_attributes": True}


class AssetInfo(BaseModel):
    """
    Basic information about an asset, returned by list operations.
    
    This is a lightweight model for discovery/listing purposes.
    For full details, use AssetMetadata.
    """
    identifier: str = Field(..., description="Unique identifier (FQN or path)")
    name: str = Field(..., description="Display name")
    asset_type: UnifiedAssetType = Field(..., description="Type of the asset")
    connector_type: str = Field(..., description="Connector type that manages this asset")
    
    # Optional quick info
    description: Optional[str] = Field(None, description="Short description")
    path: Optional[str] = Field(None, description="Full path to the asset")
    url: Optional[str] = Field(None, description="URL to view the asset in its native UI")
    
    # Container hierarchy (if applicable)
    catalog: Optional[str] = Field(None, description="Catalog/database name")
    schema_name: Optional[str] = Field(None, alias="schema", description="Schema name")
    
    model_config = {"from_attributes": True, "populate_by_name": True}
    
    @property
    def category(self) -> AssetCategory:
        """Get the category of this asset."""
        return get_asset_category(self.asset_type)
    
    @property
    def has_schema(self) -> bool:
        """Check if this asset type supports schema."""
        return supports_schema(self.asset_type)


class AssetMetadata(BaseModel):
    """
    Full metadata for an asset.
    
    This is the comprehensive model returned by get_asset_metadata operations,
    including schema, statistics, ownership, and platform-specific properties.
    """
    # Core identity
    identifier: str = Field(..., description="Unique identifier (FQN or path)")
    name: str = Field(..., description="Display name")
    asset_type: UnifiedAssetType = Field(..., description="Type of the asset")
    connector_type: str = Field(..., description="Connector type that manages this asset")
    
    # Description and documentation
    description: Optional[str] = Field(None, description="Full description")
    comment: Optional[str] = Field(None, description="Comment (if different from description)")
    
    # Location
    path: Optional[str] = Field(None, description="Full path to the asset")
    url: Optional[str] = Field(None, description="URL to view the asset in its native UI")
    location: Optional[str] = Field(None, description="Physical location (e.g., S3 path)")
    
    # Container hierarchy
    catalog: Optional[str] = Field(None, description="Catalog/database name")
    schema_name: Optional[str] = Field(None, alias="schema", description="Schema name")
    
    # Schema (for data assets)
    schema_info: Optional[SchemaInfo] = Field(None, description="Schema information")
    
    # Ownership
    ownership: Optional[AssetOwnership] = Field(None, description="Ownership information")
    
    # Statistics
    statistics: Optional[AssetStatistics] = Field(None, description="Asset statistics")
    
    # Tags and properties
    tags: Dict[str, str] = Field(default_factory=dict, description="Tags/labels")
    properties: Dict[str, Any] = Field(default_factory=dict, description="Platform-specific properties")
    
    # Lifecycle
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    created_by: Optional[str] = Field(None, description="Creator")
    updated_by: Optional[str] = Field(None, description="Last updater")
    
    # Status
    status: Optional[str] = Field(None, description="Asset status if applicable")
    
    # Validation result (populated by connector)
    exists: bool = Field(True, description="Whether the asset currently exists")
    validation_message: Optional[str] = Field(None, description="Validation message if any")
    
    model_config = {"from_attributes": True, "populate_by_name": True}
    
    @property
    def category(self) -> AssetCategory:
        """Get the category of this asset."""
        return get_asset_category(self.asset_type)
    
    @property
    def has_schema(self) -> bool:
        """Check if this asset type supports schema."""
        return supports_schema(self.asset_type)
    
    @property
    def column_count(self) -> int:
        """Get the number of columns if schema is available."""
        if self.schema_info:
            return self.schema_info.column_count
        return 0


# ============================================================================
# Validation Result
# ============================================================================

class AssetValidationResult(BaseModel):
    """Result of validating an asset's existence."""
    identifier: str = Field(..., description="Asset identifier that was validated")
    exists: bool = Field(..., description="Whether the asset exists")
    validated: bool = Field(True, description="Whether validation was performed")
    asset_type: Optional[UnifiedAssetType] = Field(None, description="Detected asset type")
    message: Optional[str] = Field(None, description="Validation message")
    details: Dict[str, Any] = Field(default_factory=dict, description="Additional details")
    
    model_config = {"from_attributes": True}


# ============================================================================
# Sample Data
# ============================================================================

class SampleData(BaseModel):
    """Sample data from an asset."""
    columns: List[str] = Field(default_factory=list, description="Column names")
    rows: List[List[Any]] = Field(default_factory=list, description="Data rows")
    total_rows: Optional[int] = Field(None, description="Total row count in the asset")
    sample_size: int = Field(0, description="Number of rows in this sample")
    truncated: bool = Field(False, description="Whether the sample was truncated")
    
    model_config = {"from_attributes": True}
    
    @property
    def row_count(self) -> int:
        """Return the number of sample rows."""
        return len(self.rows)

File: ontos/src/frontend/src/types/assets.ts
/**
 * Unified Asset Types
 * 
 * This module provides platform-agnostic asset type definitions for the
 * pluggable connector architecture. It enables Datasets and Contracts to
 * reference any asset type uniformly across Unity Catalog, Snowflake,
 * Kafka, PowerBI, and other platforms.
 * 
 * These types mirror the backend definitions in src/backend/src/models/assets.py
 */

// ============================================================================
// Asset Categories
// ============================================================================

/**
 * High-level categorization of assets by their primary function.
 */
export enum AssetCategory {
    DATA = "data",           // Tables, views, topics, streams
    COMPUTE = "compute",     // Functions, models, jobs, notebooks
    SEMANTIC = "semantic",   // Metrics, measures, dimensions, KPIs
    VISUALIZATION = "viz",   // Dashboards, reports, charts
    STORAGE = "storage",     // Files, objects, volumes
    OTHER = "other",         // Catch-all for uncategorized assets
}

// ============================================================================
// Unified Asset Types
// ============================================================================

/**
 * Unified enumeration of all supported asset types across platforms.
 * 
 * Naming convention: {PLATFORM}_{ASSET_TYPE}
 * This allows platform-specific handling while maintaining a unified interface.
 */
export enum UnifiedAssetType {
    // -------------------------------------------------------------------------
    // Unity Catalog / Databricks - Data Assets
    // -------------------------------------------------------------------------
    UC_TABLE = "uc_table",
    UC_VIEW = "uc_view",
    UC_MATERIALIZED_VIEW = "uc_materialized_view",
    UC_STREAMING_TABLE = "uc_streaming_table",
    UC_VOLUME = "uc_volume",
    UC_VOLUME_FILE = "uc_volume_file",
    
    // -------------------------------------------------------------------------
    // Unity Catalog / Databricks - Compute Assets
    // -------------------------------------------------------------------------
    UC_FUNCTION = "uc_function",
    UC_MODEL = "uc_model",
    UC_NOTEBOOK = "uc_notebook",
    UC_JOB = "uc_job",
    UC_PIPELINE = "uc_pipeline", // DLT / Spark Declarative Pipelines (SDP)
    
    // -------------------------------------------------------------------------
    // Unity Catalog / Databricks - Semantic Assets
    // -------------------------------------------------------------------------
    UC_METRIC = "uc_metric", // Unity Catalog metrics (AI/BI)
    
    // -------------------------------------------------------------------------
    // Snowflake - Data Assets
    // -------------------------------------------------------------------------
    SNOWFLAKE_TABLE = "snowflake_table",
    SNOWFLAKE_VIEW = "snowflake_view",
    SNOWFLAKE_MATERIALIZED_VIEW = "snowflake_materialized_view",
    SNOWFLAKE_STREAM = "snowflake_stream",
    SNOWFLAKE_STAGE = "snowflake_stage",
    
    // -------------------------------------------------------------------------
    // Snowflake - Compute Assets
    // -------------------------------------------------------------------------
    SNOWFLAKE_FUNCTION = "snowflake_function",
    SNOWFLAKE_PROCEDURE = "snowflake_procedure",
    SNOWFLAKE_TASK = "snowflake_task",
    
    // -------------------------------------------------------------------------
    // Kafka - Data Assets
    // -------------------------------------------------------------------------
    KAFKA_TOPIC = "kafka_topic",
    KAFKA_SCHEMA = "kafka_schema", // Schema Registry schemas
    
    // -------------------------------------------------------------------------
    // PowerBI - Visualization & Semantic Assets
    // -------------------------------------------------------------------------
    POWERBI_DATASET = "powerbi_dataset",
    POWERBI_SEMANTIC_MODEL = "powerbi_semantic_model",
    POWERBI_DASHBOARD = "powerbi_dashboard",
    POWERBI_REPORT = "powerbi_report",
    POWERBI_DATAFLOW = "powerbi_dataflow",
    
    // -------------------------------------------------------------------------
    // Cloud Storage - Storage Assets
    // -------------------------------------------------------------------------
    S3_BUCKET = "s3_bucket",
    S3_OBJECT = "s3_object",
    ADLS_CONTAINER = "adls_container",
    ADLS_PATH = "adls_path",
    GCS_BUCKET = "gcs_bucket",
    GCS_OBJECT = "gcs_object",
    
    // -------------------------------------------------------------------------
    // API & External - Other Assets
    // -------------------------------------------------------------------------
    EXTERNAL_API = "external_api",
    EXTERNAL_DATABASE = "external_database",
    
    // -------------------------------------------------------------------------
    // Generic / Unknown
    // -------------------------------------------------------------------------
    GENERIC = "generic",
    UNKNOWN = "unknown",
}

// ============================================================================
// Asset Type Metadata
// ============================================================================

/**
 * Mapping of asset types to their categories.
 */
export const ASSET_TYPE_CATEGORIES: Record<UnifiedAssetType, AssetCategory> = {
    // UC Data
    [UnifiedAssetType.UC_TABLE]: AssetCategory.DATA,
    [UnifiedAssetType.UC_VIEW]: AssetCategory.DATA,
    [UnifiedAssetType.UC_MATERIALIZED_VIEW]: AssetCategory.DATA,
    [UnifiedAssetType.UC_STREAMING_TABLE]: AssetCategory.DATA,
    [UnifiedAssetType.UC_VOLUME]: AssetCategory.STORAGE,
    [UnifiedAssetType.UC_VOLUME_FILE]: AssetCategory.STORAGE,
    // UC Compute
    [UnifiedAssetType.UC_FUNCTION]: AssetCategory.COMPUTE,
    [UnifiedAssetType.UC_MODEL]: AssetCategory.COMPUTE,
    [UnifiedAssetType.UC_NOTEBOOK]: AssetCategory.COMPUTE,
    [UnifiedAssetType.UC_JOB]: AssetCategory.COMPUTE,
    [UnifiedAssetType.UC_PIPELINE]: AssetCategory.COMPUTE,
    // UC Semantic
    [UnifiedAssetType.UC_METRIC]: AssetCategory.SEMANTIC,
    // Snowflake Data
    [UnifiedAssetType.SNOWFLAKE_TABLE]: AssetCategory.DATA,
    [UnifiedAssetType.SNOWFLAKE_VIEW]: AssetCategory.DATA,
    [UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW]: AssetCategory.DATA,
    [UnifiedAssetType.SNOWFLAKE_STREAM]: AssetCategory.DATA,
    [UnifiedAssetType.SNOWFLAKE_STAGE]: AssetCategory.STORAGE,
    // Snowflake Compute
    [UnifiedAssetType.SNOWFLAKE_FUNCTION]: AssetCategory.COMPUTE,
    [UnifiedAssetType.SNOWFLAKE_PROCEDURE]: AssetCategory.COMPUTE,
    [UnifiedAssetType.SNOWFLAKE_TASK]: AssetCategory.COMPUTE,
    // Kafka
    [UnifiedAssetType.KAFKA_TOPIC]: AssetCategory.DATA,
    [UnifiedAssetType.KAFKA_SCHEMA]: AssetCategory.DATA,
    // PowerBI
    [UnifiedAssetType.POWERBI_DATASET]: AssetCategory.DATA,
    [UnifiedAssetType.POWERBI_SEMANTIC_MODEL]: AssetCategory.SEMANTIC,
    [UnifiedAssetType.POWERBI_DASHBOARD]: AssetCategory.VISUALIZATION,
    [UnifiedAssetType.POWERBI_REPORT]: AssetCategory.VISUALIZATION,
    [UnifiedAssetType.POWERBI_DATAFLOW]: AssetCategory.DATA,
    // Cloud Storage
    [UnifiedAssetType.S3_BUCKET]: AssetCategory.STORAGE,
    [UnifiedAssetType.S3_OBJECT]: AssetCategory.STORAGE,
    [UnifiedAssetType.ADLS_CONTAINER]: AssetCategory.STORAGE,
    [UnifiedAssetType.ADLS_PATH]: AssetCategory.STORAGE,
    [UnifiedAssetType.GCS_BUCKET]: AssetCategory.STORAGE,
    [UnifiedAssetType.GCS_OBJECT]: AssetCategory.STORAGE,
    // External
    [UnifiedAssetType.EXTERNAL_API]: AssetCategory.OTHER,
    [UnifiedAssetType.EXTERNAL_DATABASE]: AssetCategory.DATA,
    // Generic
    [UnifiedAssetType.GENERIC]: AssetCategory.OTHER,
    [UnifiedAssetType.UNKNOWN]: AssetCategory.OTHER,
};

/**
 * Asset types that support schema (columns/fields).
 */
export const SCHEMA_SUPPORTING_TYPES = new Set<UnifiedAssetType>([
    UnifiedAssetType.UC_TABLE,
    UnifiedAssetType.UC_VIEW,
    UnifiedAssetType.UC_MATERIALIZED_VIEW,
    UnifiedAssetType.UC_STREAMING_TABLE,
    UnifiedAssetType.SNOWFLAKE_TABLE,
    UnifiedAssetType.SNOWFLAKE_VIEW,
    UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW,
    UnifiedAssetType.KAFKA_TOPIC, // If using Schema Registry
    UnifiedAssetType.POWERBI_DATASET,
    UnifiedAssetType.POWERBI_SEMANTIC_MODEL,
]);

/**
 * Connector types for different platforms.
 */
export type ConnectorType = "databricks" | "snowflake" | "kafka" | "powerbi" | "s3" | "custom";

/**
 * Mapping of connector types to their supported asset types.
 */
export const CONNECTOR_ASSET_TYPES: Record<ConnectorType, UnifiedAssetType[]> = {
    databricks: [
        UnifiedAssetType.UC_TABLE,
        UnifiedAssetType.UC_VIEW,
        UnifiedAssetType.UC_MATERIALIZED_VIEW,
        UnifiedAssetType.UC_STREAMING_TABLE,
        UnifiedAssetType.UC_VOLUME,
        UnifiedAssetType.UC_VOLUME_FILE,
        UnifiedAssetType.UC_FUNCTION,
        UnifiedAssetType.UC_MODEL,
        UnifiedAssetType.UC_NOTEBOOK,
        UnifiedAssetType.UC_JOB,
        UnifiedAssetType.UC_PIPELINE,
        UnifiedAssetType.UC_METRIC,
    ],
    snowflake: [
        UnifiedAssetType.SNOWFLAKE_TABLE,
        UnifiedAssetType.SNOWFLAKE_VIEW,
        UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW,
        UnifiedAssetType.SNOWFLAKE_STREAM,
        UnifiedAssetType.SNOWFLAKE_STAGE,
        UnifiedAssetType.SNOWFLAKE_FUNCTION,
        UnifiedAssetType.SNOWFLAKE_PROCEDURE,
        UnifiedAssetType.SNOWFLAKE_TASK,
    ],
    kafka: [
        UnifiedAssetType.KAFKA_TOPIC,
        UnifiedAssetType.KAFKA_SCHEMA,
    ],
    powerbi: [
        UnifiedAssetType.POWERBI_DATASET,
        UnifiedAssetType.POWERBI_SEMANTIC_MODEL,
        UnifiedAssetType.POWERBI_DASHBOARD,
        UnifiedAssetType.POWERBI_REPORT,
        UnifiedAssetType.POWERBI_DATAFLOW,
    ],
    s3: [
        UnifiedAssetType.S3_BUCKET,
        UnifiedAssetType.S3_OBJECT,
    ],
    custom: [
        UnifiedAssetType.GENERIC,
        UnifiedAssetType.EXTERNAL_API,
        UnifiedAssetType.EXTERNAL_DATABASE,
    ],
};

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Get the category for a given asset type.
 */
export function getAssetCategory(assetType: UnifiedAssetType): AssetCategory {
    return ASSET_TYPE_CATEGORIES[assetType] || AssetCategory.OTHER;
}

/**
 * Check if an asset type supports schema (columns/fields).
 */
export function supportsSchema(assetType: UnifiedAssetType): boolean {
    return SCHEMA_SUPPORTING_TYPES.has(assetType);
}

/**
 * Get the connector type for a given asset type.
 */
export function getConnectorTypeForAsset(assetType: UnifiedAssetType): ConnectorType | null {
    for (const [connectorType, assetTypes] of Object.entries(CONNECTOR_ASSET_TYPES)) {
        if (assetTypes.includes(assetType)) {
            return connectorType as ConnectorType;
        }
    }
    return null;
}

/**
 * Get a display name for an asset type.
 */
export function getAssetTypeDisplayName(assetType: UnifiedAssetType): string {
    const displayNames: Record<UnifiedAssetType, string> = {
        // Unity Catalog
        [UnifiedAssetType.UC_TABLE]: "Unity Catalog Table",
        [UnifiedAssetType.UC_VIEW]: "Unity Catalog View",
        [UnifiedAssetType.UC_MATERIALIZED_VIEW]: "Unity Catalog Materialized View",
        [UnifiedAssetType.UC_STREAMING_TABLE]: "Unity Catalog Streaming Table",
        [UnifiedAssetType.UC_VOLUME]: "Unity Catalog Volume",
        [UnifiedAssetType.UC_VOLUME_FILE]: "Unity Catalog Volume File",
        [UnifiedAssetType.UC_FUNCTION]: "Unity Catalog Function",
        [UnifiedAssetType.UC_MODEL]: "Unity Catalog Model",
        [UnifiedAssetType.UC_NOTEBOOK]: "Notebook",
        [UnifiedAssetType.UC_JOB]: "Job",
        [UnifiedAssetType.UC_PIPELINE]: "Declarative Pipeline (DLT/SDP)",
        [UnifiedAssetType.UC_METRIC]: "Unity Catalog Metric",
        // Snowflake
        [UnifiedAssetType.SNOWFLAKE_TABLE]: "Snowflake Table",
        [UnifiedAssetType.SNOWFLAKE_VIEW]: "Snowflake View",
        [UnifiedAssetType.SNOWFLAKE_MATERIALIZED_VIEW]: "Snowflake MV",
        [UnifiedAssetType.SNOWFLAKE_STREAM]: "Snowflake Stream",
        [UnifiedAssetType.SNOWFLAKE_STAGE]: "Snowflake Stage",
        [UnifiedAssetType.SNOWFLAKE_FUNCTION]: "Snowflake Function",
        [UnifiedAssetType.SNOWFLAKE_PROCEDURE]: "Snowflake Procedure",
        [UnifiedAssetType.SNOWFLAKE_TASK]: "Snowflake Task",
        // Kafka
        [UnifiedAssetType.KAFKA_TOPIC]: "Kafka Topic",
        [UnifiedAssetType.KAFKA_SCHEMA]: "Kafka Schema",
        // PowerBI
        [UnifiedAssetType.POWERBI_DATASET]: "Power BI Dataset",
        [UnifiedAssetType.POWERBI_SEMANTIC_MODEL]: "Power BI Semantic Model",
        [UnifiedAssetType.POWERBI_DASHBOARD]: "Power BI Dashboard",
        [UnifiedAssetType.POWERBI_REPORT]: "Power BI Report",
        [UnifiedAssetType.POWERBI_DATAFLOW]: "Power BI Dataflow",
        // Storage
        [UnifiedAssetType.S3_BUCKET]: "S3 Bucket",
        [UnifiedAssetType.S3_OBJECT]: "S3 Object",
        [UnifiedAssetType.ADLS_CONTAINER]: "ADLS Container",
        [UnifiedAssetType.ADLS_PATH]: "ADLS Path",
        [UnifiedAssetType.GCS_BUCKET]: "GCS Bucket",
        [UnifiedAssetType.GCS_OBJECT]: "GCS Object",
        // External
        [UnifiedAssetType.EXTERNAL_API]: "External API",
        [UnifiedAssetType.EXTERNAL_DATABASE]: "External Database",
        // Generic
        [UnifiedAssetType.GENERIC]: "Generic",
        [UnifiedAssetType.UNKNOWN]: "Unknown",
    };
    return displayNames[assetType] || assetType;
}

/**
 * Get an icon name for an asset type (for use with icon libraries).
 */
export function getAssetTypeIcon(assetType: UnifiedAssetType): string {
    const iconMap: Partial<Record<UnifiedAssetType, string>> = {
        // UC
        [UnifiedAssetType.UC_TABLE]: "table",
        [UnifiedAssetType.UC_VIEW]: "eye",
        [UnifiedAssetType.UC_MATERIALIZED_VIEW]: "layers",
        [UnifiedAssetType.UC_STREAMING_TABLE]: "activity",
        [UnifiedAssetType.UC_VOLUME]: "hard-drive",
        [UnifiedAssetType.UC_FUNCTION]: "code",
        [UnifiedAssetType.UC_MODEL]: "brain",
        [UnifiedAssetType.UC_NOTEBOOK]: "file-text",
        [UnifiedAssetType.UC_JOB]: "play-circle",
        [UnifiedAssetType.UC_METRIC]: "bar-chart-2",
        // Snowflake
        [UnifiedAssetType.SNOWFLAKE_TABLE]: "table",
        [UnifiedAssetType.SNOWFLAKE_VIEW]: "eye",
        // Kafka
        [UnifiedAssetType.KAFKA_TOPIC]: "message-square",
        // PowerBI
        [UnifiedAssetType.POWERBI_DASHBOARD]: "layout-dashboard",
        [UnifiedAssetType.POWERBI_REPORT]: "file-bar-chart",
        [UnifiedAssetType.POWERBI_SEMANTIC_MODEL]: "box",
    };
    return iconMap[assetType] || "file";
}

// ============================================================================
// Asset Info Interfaces
// ============================================================================

/**
 * Basic information about an asset, returned by list operations.
 */
export interface AssetInfo {
    identifier: string;
    name: string;
    asset_type: UnifiedAssetType;
    connector_type: ConnectorType;
    description?: string | null;
    path?: string | null;
    url?: string | null;
    catalog?: string | null;
    schema?: string | null;
}

/**
 * Column information for assets with schema.
 */
export interface ColumnInfo {
    name: string;
    data_type: string;
    logical_type?: string | null;
    nullable: boolean;
    description?: string | null;
    is_primary_key: boolean;
    is_partition_key: boolean;
    default_value?: string | null;
    tags: Record<string, string>;
    properties: Record<string, unknown>;
}

/**
 * Schema information for assets that have columnar structure.
 */
export interface SchemaInfo {
    columns: ColumnInfo[];
    primary_key?: string[] | null;
    partition_columns?: string[] | null;
    clustering_columns?: string[] | null;
    schema_version?: string | null;
}

/**
 * Ownership information for an asset.
 */
export interface AssetOwnership {
    owner?: string | null;
    owner_email?: string | null;
    steward?: string | null;
    team?: string | null;
}

/**
 * Statistics about an asset.
 */
export interface AssetStatistics {
    row_count?: number | null;
    size_bytes?: number | null;
    partition_count?: number | null;
    file_count?: number | null;
    last_modified?: string | null;
    last_accessed?: string | null;
}

/**
 * Full metadata for an asset.
 */
export interface AssetMetadata extends AssetInfo {
    comment?: string | null;
    location?: string | null;
    schema_info?: SchemaInfo | null;
    ownership?: AssetOwnership | null;
    statistics?: AssetStatistics | null;
    tags: Record<string, string>;
    properties: Record<string, unknown>;
    created_at?: string | null;
    updated_at?: string | null;
    created_by?: string | null;
    updated_by?: string | null;
    status?: string | null;
    exists: boolean;
    validation_message?: string | null;
}

/**
 * Result of validating an asset's existence.
 */
export interface AssetValidationResult {
    identifier: string;
    exists: boolean;
    validated: boolean;
    asset_type?: UnifiedAssetType | null;
    message?: string | null;
    details: Record<string, unknown>;
}

/**
 * Sample data from an asset.
 */
export interface SampleData {
    columns: string[];
    rows: unknown[][];
    total_rows?: number | null;
    sample_size: number;
    truncated: boolean;
}

